subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/jar/JarClasspathSnapshotFactory.java

```diff
@@ -50,7 +61,45 @@ JarClasspathSnapshot createSnapshot(Iterable<JarArchive> jarArchives) {
                 }
             }
         }
+
         JarClasspathSnapshotData jarClasspathSnapshotData = new JarClasspathSnapshotData(jarHashes, duplicateClasses);
         return new JarClasspathSnapshot(jarSnapshots, jarClasspathSnapshotData);
     }
+
+    private Set<CreateJarSnapshot> snapshotAll(final Iterable<JarArchive> jarArchives) {
+        final Set<CreateJarSnapshot> snapshotOperations = Sets.newLinkedHashSet();
+
+        buildOperationExecutor.runAll(new Action<BuildOperationQueue<CreateJarSnapshot>>() {
+            @Override
+            public void execute(BuildOperationQueue<CreateJarSnapshot> buildOperationQueue) {
+                for (JarArchive jar : jarArchives) {
+                    CreateJarSnapshot operation = new CreateJarSnapshot(jar);
```

Comment:
How about moving the check for `jar.file.exists()` here and only instantiate `CreateJarSnapshot` if so? Or is that bit worth executing in parallel, too?

Reply:
It's IO so it's probably worth doing in parallel.